### PR TITLE
[Path] Refactor path array algorithm into independent `PathArray` class

### PR DESCRIPTION
**Pull request by Russ4262**
New class is `obj` independent allowing for greater scripting capability.
Added missing `onDocumentRestored()` method to ObjectArray class to handle property visibility and edit modes, per standard Path operations on document restoration.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

File: src/Mod/Path/PathScripts/PathArray.py [hunk 1]
```diff
@@ -75,39 +75,35 @@ def __setstate__(self, state):
...
+            Angle = Centre = CopiesX = CopiesY = SwapDirection = 2
+            Copies = Offset = 0
+        elif obj.Type == 'Linear2D':
+            Angle = Copies = Centre = 2
+            CopiesX = CopiesY = Offset = SwapDirection = 0
+        elif obj.Type == 'Polar':
+            Angle = Copies = Centre = 0
+            CopiesX = CopiesY = Offset = SwapDirection = 2
+
+        obj.setEditorMode('Angle', Angle)
```
Comment (sliptonic) [inline src/Mod/Path/PathScripts/PathArray.py:88]: The variable names are ambiguous.  What's being set is the editor mode, not the angle/copy count ,etc.

Comment (Russ4262) [reply src/Mod/Path/PathScripts/PathArray.py:88]: I will gladly revert to existing code if you like.

Comment (sliptonic) [reply src/Mod/Path/PathScripts/PathArray.py:88]: The code is fine. It's more compact than the original which I appreciate.  But the variable name is confusing. When I see a variable name like 'Angle', I expect it to contain...a value that will be interpreted as an angle.  But in this case the variable 'Angle' contains a value that will be interpreted as an EditorMode.  So the variable name should be something like angleControlMode or angleControlVisibility.   Variable names is one place to _not_ be compact.  Be explicit.

Comment (Russ4262) [reply src/Mod/Path/PathScripts/PathArray.py]: Variable names have been improved per indicated expectations.  Currently, 1 July 2021, builds are failing due to FEM unit tests.

Thanks for the feedback and oversight.

**Pull request merged**